tests/components/humidifier/test_reproduce_state.py

```diff
@@ -0,0 +1,237 @@
+"""The tests for reproduction of state."""
+
+import pytest
+
+from homeassistant.components.humidifier.const import (
+    ATTR_HUMIDITY,
+    ATTR_MODE,
+    DOMAIN,
+    MODE_AWAY,
+    MODE_ECO,
+    MODE_NORMAL,
+    SERVICE_SET_HUMIDITY,
+    SERVICE_SET_MODE,
+)
+from homeassistant.components.humidifier.reproduce_state import async_reproduce_states
+from homeassistant.const import SERVICE_TURN_OFF, SERVICE_TURN_ON, STATE_OFF, STATE_ON
+from homeassistant.core import Context, State
+
+from tests.common import async_mock_service
+
+ENTITY_1 = "humidifier.test1"
+ENTITY_2 = "humidifier.test2"
+
+
+async def test_reproducing_on_off_states(hass, caplog):
+    """Test reproducing humidifier states."""
+    hass.states.async_set(ENTITY_1, "off", {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 45})
+    hass.states.async_set(ENTITY_2, "on", {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 45})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    mode_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_MODE)
+    humidity_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_HUMIDITY)
+
+    # These calls should do nothing as entities already in desired state
+    await hass.helpers.state.async_reproduce_state(
+        [
+            State(ENTITY_1, "off", {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 45}),
+            State(ENTITY_2, "on", {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 45}),
+        ],
+    )
+
+    assert len(turn_on_calls) == 0
+    assert len(turn_off_calls) == 0
+    assert len(mode_calls) == 0
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5745133)